### DISCUSSION TITLE
how to using mocha for testing

### DISCUSSION THREAD

**Discussion by morriswinkler** [nrwl/nx#10528] [Show and Tell]
I added my mocha tests with the `nx:run-commands` executor.

Is there any other way of doing it?

**Comment by tomerguttman** [comment]
up

**Comment by yonatan-jux** [comment]
@morriswinkler 
could you please share how you did it with the `nx:run-commands`?

**Comment by morriswinkler** [comment]
```json
{
    "$schema": "../../node_modules/nx/schemas/project-schema.json",
    "sourceRoot": "libs/my-package/",
    "projectType": "library",
    "targets": {
        "test": {
            "executor": "nx:run-commands",
            "dependsOn": [{"projects": "self", "target": "build"}],
            "options": {
                "commands": [
                    "mocha"
                ]
            }
       }
   }
}
```

**Reply by lakshmikanthadn** [reply]
The above solution works great for the JS application without any local package dependency.

But How do we configure this for a Typescript application that has local package dependencies?

My first solution was,
Build the application along with its dependent packages then run the test,
**but I get an error that says local package not found**, 

Any help is really appreciated. we are migrating to NX.

**Comment by eXigentCoder** [comment]
I got mocha working in my nx repo with local package dependencies using a script in the `package.json` file, though I'm sure it could be easily updated to run in the executor. The command that worked for me was:

```"mocha": "cross-env NODE_ENV=test mocha src/**/*.test.ts --require ts-node/register -r tsconfig-paths/register --recursive --extension .ts --exit",```

Hope that helps!

**Comment by wernc23** [comment]
Hey all, I was able to get it working with the following

```   
 "test": {
      "executor": "nx:run-commands",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "dependsOn": [{ "projects": "self", "target": "build" }],
      "options": {
        "commands": [
          "nyc --reporter=lcov --reporter=text --reporter=html mocha --config .mocharc.js --require ts-node/register --require source-map-support/register --recursive --exclude \"node_modules/**/*\" \"**/*spec.ts\""
        ],
        "cwd": "libs/my-library"
      }
    }
    
```
    

   The `cwd` was important here. So from the root I can run `npx nx test my-library` and it now runs using mocha. My mocharc is straight forward

```   
   module.exports = {
  exclude: ['node_modules/**/*'],
  extension: ['ts'],
  forbidOnly: true,
  forbidPending: true,
  require: ['ts-node/register', 'source-map-support/register'],
  spec: ['**/*.spec.ts'],
};
```

Hope this helps someone.
